src/index.js

```diff
@@ -115,20 +115,54 @@ const Keys = {
   Backspace: { keyCode: 8, code: "Backspace", key: "Backspace" }
 };
 
+/**
+ * Method to abstract setting the value of an element
+ *
+ * We need to override the logic here because React will wrap an `input`'s
+ * `.value` setter and prevent us from dispatching `onChange` events after
+ * setting the value with `value = something`. Here, we'll call the prototype
+ * value setter method so the native `.value` will be correct, fire the event
+ * (the event.target will have the correct value, as will the original `input`),
+ * and then use the prototype override of `value` if there is one for React's
+ * internals.
+ *
+ * @param {Element} element Element we want to set the `value` on
+ * @param {String} value Value we want to set on `element`
+ */
+function setElementValue(element, value) {
+  const valueSetter = Object.getOwnPropertyDescriptor(element, "value").set;
+  const prototypeValueSetter = Object.getOwnPropertyDescriptor(
+    Object.getPrototypeOf(element),
+    "value"
+  ).set;
+
+  // Call the native prototype method
+  prototypeValueSetter.call(element, value);
+
+  fireEvent.change(element, { bubbles: true });
+
+  // If the prototype method is not the same as the method on the element, then
+  // there's an override we need to also call.
+  if (valueSetter !== prototypeValueSetter) {
+    element.value = value;
+  }
+}
+
 function backspace(element) {
-  const eventOptions = {
+  const keyboardEventOptions = {
     key: Keys.Backspace.key,
     keyCode: Keys.Backspace.keyCode,
     which: Keys.Backspace.keyCode
   };
-  fireEvent.keyDown(element, eventOptions);
-  fireEvent.keyUp(element, eventOptions);
+  fireEvent.keyDown(element, keyboardEventOptions);
+  fireEvent.keyUp(element, keyboardEventOptions);
 
   if (!element.readOnly) {
     fireEvent.input(element, {
       inputType: "deleteContentBackward"
     });
-    element.value = ""; // when we add special keys to API, will need to respect selected range
+
+    setElementValue(element, "");
```

Comment:
Hmmm... Is there a reason we can't just use `fireEvent.change(element, {target: {value}})`? The `setElementValue` function is basically what `fireEvent.change` does: https://github.com/testing-library/dom-testing-library/blob/8846eaf20972f8e41ed11f278948ac38a692c3f1/src/events.js#L81-L96

Reply:
Great question. I tried to explain this in the PR description but I obviously struggled and didn't get the point across. Also, I hadn't yet seen the `setNativeValue` implementation. Thanks for the patience of letting me try again.

My first attempt tried to use `fireEvent.change(element, { target: { value: '' }})` without realizing that they have a very similar function to the function that you linked to (https://github.com/testing-library/dom-testing-library/blob/8846eaf20972f8e41ed11f278948ac38a692c3f1/src/events.js#L81-L96) above https://github.com/testing-library/user-event/pull/286#discussion_r427617114.

If I replace this `setElementValue( ... )` with the `fireEvent.change`, the problem is that React's internal representation of state will fall out of sync with the actual `input.value` because React internally caches the `input` value https://github.com/facebook/react/blob/8b3b5c35245b3aa0b2802b51856fe9dad5bcb671/packages/react-dom/src/client/inputValueTracking.js#L51-L104. When we set the native `.value`, the React `value` setter isn't called, causing React to have a different internal value than the `input` itself. Fortunately the test additions I added will catch this. Changing the value will show that the value is not maintained.

This is the part of my abstraction that differs from the above linked: https://github.com/testing-library/user-event/pull/286/files#diff-1fdf421c05c1140f6d71444ea2b27638R144-R148, which calls the prototype overridden value setter after the event is dispatched.

However, there's a way simpler solution that I didn't see and is now obvious after looking at the `fireEvent.change` and these new tests. If I just call `input.value = ""` after calling `fireEvent.change` then everything works as we'd expect! I'll make this change.